File: internal/handlers/group_handler.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type GroupHandler struct {
	db *database.Database
}

func NewGroupHandler(db *database.Database) *GroupHandler {
	return &GroupHandler{db: db}
}

type CreateGroupRequest struct {
	GroupName string   `json:"group_name" binding:"required"`
	GroupIcon string   `json:"group_icon,omitempty"`
	MemberIDs []string `json:"member_ids"`
}

func (h *GroupHandler) CreateGroup(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req CreateGroupRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	members := []primitive.ObjectID{userIDObj}
	for _, memberIDStr := range req.MemberIDs {
		memberID, err := primitive.ObjectIDFromHex(memberIDStr)
		if err != nil {
			continue
		}
		members = append(members, memberID)
	}

	chat := models.Chat{
		ID:        primitive.NewObjectID(),
		Type:      "group",
		Members:   members,
		Admins: []models.AdminRole{
			{
				UserID:      userIDObj,
				Role:        "owner",
				Permissions: []string{"all"},
				GrantedAt:   time.Now(),
				GrantedBy:   userIDObj,
			},
		},
		GroupName: req.GroupName,
		GroupIcon: req.GroupIcon,
		MaxMembers: 200000, // Telegram limit
		Statistics: models.ChatStatistics{
			LastCalculated: time.Now(),
		},
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	_, err := h.db.MongoDB.Collection("chats").InsertOne(context.Background(), chat)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create group"})
		return
	}

	c.JSON(http.StatusCreated, chat)
}

func (h *GroupHandler) GetGroups(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	cursor, err := h.db.MongoDB.Collection("chats").Find(
		context.Background(),
		bson.M{
			"type":    "group",
			"members": userIDObj,
		},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch groups"})
		return
	}
	defer cursor.Close(context.Background())

	var groups []models.Chat
	if err := cursor.All(context.Background(), &groups); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode groups"})
		return
	}

	c.JSON(http.StatusOK, groups)
}

func (h *GroupHandler) GetGroup(c *gin.Context) {
	groupIDStr := c.Param("group_id")
	groupID, err := primitive.ObjectIDFromHex(groupIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid group ID"})
		return
	}

	var group models.Chat
	err = h.db.MongoDB.Collection("chats").FindOne(
		context.Background(),
		bson.M{"_id": groupID, "type": "group"},
	).Decode(&group)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Group not found"})
		return
	}

	c.JSON(http.StatusOK, group)
}

func (h *GroupHandler) UpdateGroup(c *gin.Context) {
	groupIDStr := c.Param("group_id")
	groupID, err := primitive.ObjectIDFromHex(groupIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid group ID"})
		return
	}

	var updateData map[string]interface{}
	if err := c.ShouldBindJSON(&updateData); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	updateData["updated_at"] = time.Now()
	_, err = h.db.MongoDB.Collection("chats").UpdateOne(
		context.Background(),
		bson.M{"_id": groupID, "type": "group"},
		bson.M{"$set": updateData},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update group"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Group updated successfully"})
}

func (h *GroupHandler) DeleteGroup(c *gin.Context) {
	groupIDStr := c.Param("group_id")
	groupID, err := primitive.ObjectIDFromHex(groupIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid group ID"})
		return
	}

	_, err = h.db.MongoDB.Collection("chats").DeleteOne(
		context.Background(),
		bson.M{"_id": groupID, "type": "group"},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete group"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Group deleted successfully"})
}

func (h *GroupHandler) AddMember(c *gin.Context) {
	groupIDStr := c.Param("group_id")
	groupID, err := primitive.ObjectIDFromHex(groupIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid group ID"})
		return
	}

	var req struct {
		MemberID string `json:"member_id" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	memberID, err := primitive.ObjectIDFromHex(req.MemberID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid member ID"})
		return
	}

	_, err = h.db.MongoDB.Collection("chats").UpdateOne(
		context.Background(),
		bson.M{"_id": groupID},
		bson.M{"$addToSet": bson.M{"members": memberID}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to add member"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Member added successfully"})
}

func (h *GroupHandler) RemoveMember(c *gin.Context) {
	groupIDStr := c.Param("group_id")
	groupID, err := primitive.ObjectIDFromHex(groupIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid group ID"})
		return
	}

	memberIDStr := c.Param("member_id")
	memberID, err := primitive.ObjectIDFromHex(memberIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid member ID"})
		return
	}

	_, err = h.db.MongoDB.Collection("chats").UpdateOne(
		context.Background(),
		bson.M{"_id": groupID},
		bson.M{"$pull": bson.M{"members": memberID}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to remove member"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Member removed successfully"})
}

func (h *GroupHandler) GetStatistics(c *gin.Context) {
	groupIDStr := c.Param("group_id")
	groupID, err := primitive.ObjectIDFromHex(groupIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid group ID"})
		return
	}

	var group models.Chat
	err = h.db.MongoDB.Collection("chats").FindOne(
		context.Background(),
		bson.M{"_id": groupID, "type": "group"},
	).Decode(&group)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Group not found"})
		return
	}

	// Calculate statistics
	messageCount, _ := h.db.MongoDB.Collection("messages").CountDocuments(
		context.Background(),
		bson.M{"chat_id": groupID},
	)

	mediaCount, _ := h.db.MongoDB.Collection("messages").CountDocuments(
		context.Background(),
		bson.M{
			"chat_id": groupID,
			"message_type": bson.M{"$in": []string{"image", "video", "audio", "file"}},
		},
	)

	// Update statistics
	stats := models.ChatStatistics{
		TotalMessages:  messageCount,
		TotalMedia:     mediaCount,
		ActiveMembers:  len(group.Members),
		LastCalculated: time.Now(),
	}

	_, err = h.db.MongoDB.Collection("chats").UpdateOne(
		context.Background(),
		bson.M{"_id": groupID},
		bson.M{"$set": bson.M{"statistics": stats}},
	)

	c.JSON(http.StatusOK, stats)
}

File: src/handlers/groupHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { randomUUID } = require('crypto');
const { getDB } = require('../database');
const { hub } = require('../websocket/hub');

// ─── Handlers ─────────────────────────────────────────────────────────────────

async function createGroup(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const { group_name, group_icon, member_ids } = req.body || {};

    if (!group_name) return res.status(400).json({ error: 'group_name is required' });

    const members = [userId];
    if (Array.isArray(member_ids)) {
      for (const mid of member_ids) {
        const midStr = mid.toString();
        if (!members.includes(midStr)) members.push(midStr);
      }
    }

    const now = new Date();
    const groupDoc = {
      type: 'group',
      group_name,
      group_icon: group_icon || null,
      members,
      admins: [
        {
          user_id: userId,
          role: 'owner',
          permissions: ['all'],
          granted_at: now,
          granted_by: userId,
        },
      ],
      max_members: 200000,
      statistics: {
        last_calculated: now,
      },
      created_at: now,
      updated_at: now,
    };

    const insertResult = await db.collection('chats').insertOne(groupDoc);
    groupDoc._id = insertResult.insertedId;
    const groupIdStr = insertResult.insertedId.toString();

    // Notify each member EXCEPT the creator so their sidebar shows the new group
    for (const memberId of members) {
      if (memberId !== userId) {
        hub.sendToUser(memberId.toString(), {
          type: 'new_chat',
          chat_id: groupIdStr,
          chat_type: 'group',
          group_name: group_name,
          members,
          created_by: userId,
          message_id: randomUUID(),
          timestamp: now.toISOString(),
        });
      }
    }

    return res.status(201).json({ ...groupDoc, id: groupIdStr });
  } catch (err) {
    console.error('createGroup error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getGroups(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;

    const groups = await db.collection('chats').find({
      type: 'group',
      members: userId,
    }).toArray();

    return res.json(groups.map(g => ({ ...g, id: g._id.toString() })));
  } catch (err) {
    console.error('getGroups error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getGroup(req, res) {
  try {
    const db = getDB();
    const groupIdStr = req.params.group_id;

    let groupObjId;
    try {
      groupObjId = new ObjectId(groupIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid group ID' });
    }

    const group = await db.collection('chats').findOne({ _id: groupObjId, type: 'group' });
    if (!group) return res.status(404).json({ error: 'Group not found' });

    return res.json({ ...group, id: group._id.toString() });
  } catch (err) {
    console.error('getGroup error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateGroup(req, res) {
  try {
    const db = getDB();
    const groupIdStr = req.params.group_id;
    const updates = req.body || {};

    let groupObjId;
    try {
      groupObjId = new ObjectId(groupIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid group ID' });
    }

    const setFields = { ...updates, updated_at: new Date() };
    // Remove protected fields
    delete setFields._id;
    delete setFields.type;
    delete setFields.members;
    delete setFields.admins;

    await db.collection('chats').updateOne(
      { _id: groupObjId, type: 'group' },
      { $set: setFields }
    );

    return res.json({ message: 'Group updated' });
  } catch (err) {
    console.error('updateGroup error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function deleteGroup(req, res) {
  try {
    const db = getDB();
    const groupIdStr = req.params.group_id;

    let groupObjId;
    try {
      groupObjId = new ObjectId(groupIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid group ID' });
    }

    // Fetch group BEFORE deleting so we can notify all members
    const group = await db.collection('chats').findOne({ _id: groupObjId, type: 'group' });
    if (!group) {
      return res.status(404).json({ error: 'Group not found' });
    }

    const formerMembers = Array.isArray(group.members) ? [...group.members] : [];

    const deleteResult = await db.collection('chats').deleteOne({
      _id: groupObjId,
      type: 'group',
    });

    if (deleteResult.deletedCount === 0) {
      return res.status(404).json({ error: 'Group not found' });
    }

    // Notify ALL former members (including the deleter) so they remove it from their sidebar
    const now = new Date();
    const notifyPayload = {
      type: 'group_deleted',
      chat_id: groupIdStr,
      group_name: group.group_name || null,
      message_id: randomUUID(),
      timestamp: now.toISOString(),
    };
    for (const memberId of formerMembers) {
      hub.sendToUser(memberId.toString(), notifyPayload);
    }

    return res.json({ message: 'Group deleted' });
  } catch (err) {
    console.error('deleteGroup error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function addMember(req, res) {
  try {
    const db = getDB();
    const groupIdStr = req.params.group_id;
    const { member_id } = req.body || {};

    if (!member_id) return res.status(400).json({ error: 'member_id is required' });

    let groupObjId;
    try {
      groupObjId = new ObjectId(groupIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid group ID' });
    }

    await db.collection('chats').updateOne(
      { _id: groupObjId, type: 'group' },
      {
        $addToSet: { members: member_id.toString() },
        $set: { updated_at: new Date() },
      }
    );

    return res.json({ message: 'Member added' });
  } catch (err) {
    console.error('addMember error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function removeMember(req, res) {
  try {
    const db = getDB();
    const groupIdStr = req.params.group_id;
    const memberIdStr = req.params.member_id;

    let groupObjId;
    try {
      groupObjId = new ObjectId(groupIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid group ID' });
    }

    await db.collection('chats').updateOne(
      { _id: groupObjId, type: 'group' },
      {
        $pull: { members: memberIdStr },
        $set: { updated_at: new Date() },
      }
    );

    return res.json({ message: 'Member removed' });
  } catch (err) {
    console.error('removeMember error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getStatistics(req, res) {
  try {
    const db = getDB();
    const groupIdStr = req.params.group_id;

    let groupObjId;
    try {
      groupObjId = new ObjectId(groupIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid group ID' });
    }

    const group = await db.collection('chats').findOne({ _id: groupObjId, type: 'group' });
    if (!group) return res.status(404).json({ error: 'Group not found' });

    const chatIdStr = groupObjId.toString();
    const now = new Date();

    const messageCount = await db.collection('messages').countDocuments({
      chat_id: chatIdStr,
      is_deleted: { $ne: true },
    });

    const mediaCount = await db.collection('messages').countDocuments({
      chat_id: chatIdStr,
      is_deleted: { $ne: true },
      message_type: { $in: ['image', 'video', 'audio', 'file'] },
    });

    const stats = {
      member_count: Array.isArray(group.members) ? group.members.length : 0,
      message_count: messageCount,
      media_count: mediaCount,
      last_calculated: now,
    };

    await db.collection('chats').updateOne(
      { _id: groupObjId },
      { $set: { statistics: stats, updated_at: now } }
    );

    return res.json(stats);
  } catch (err) {
    console.error('getStatistics error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

module.exports = {
  createGroup,
  getGroups,
  getGroup,
  updateGroup,
  deleteGroup,
  addMember,
  removeMember,
  getStatistics,
};
